.github/workflows/test.yml

```diff
@@ -16,104 +52,237 @@ jobs:
         run: |
           wget -nv https://github.com/riscv-collab/riscv-gnu-toolchain/releases/download/2023.05.24/riscv32-elf-ubuntu-22.04-nightly-2023.05.24-nightly.tar.gz
           tar -zxf riscv32-elf-ubuntu-22.04-nightly-2023.05.24-nightly.tar.gz
-          echo "PATH=$PATH:${GITHUB_WORKSPACE}/device/riscv/bin" >> $GITHUB_ENV
-          ls ${GITHUB_WORKSPACE}/device/riscv/bin
+
+      - name: Add RISC-V GCC to PATH
+        run: echo "PATH=$PATH:${GITHUB_WORKSPACE}/device/riscv/bin" >> $GITHUB_ENV
+
       - uses: dtolnay/rust-toolchain@stable
         with:
           targets: riscv32imc-unknown-none-elf
-          components: clippy
+
+      # Restore cargo cache
+      - uses: actions/cache@v4
+        with:
+          path: ~/.cargo/
+          key: ${{ runner.os }}-cargo-registry-${{ hashFiles('./Cargo.lock') }}-v1
+          restore-keys: |
+            ${{ runner.os }}-cargo-registry-
+
+      # Device-specific target cache
+      - uses: actions/cache@v4
+        with:
+          path: target/riscv32imc-unknown-none-elf/
+          key: ${{ runner.os }}-device-target-${{ hashFiles('device/Cargo.toml', './Cargo.lock') }}-v1
+          restore-keys: |
+            ${{ runner.os }}-device-target-
+
+      - uses: taiki-e/install-action@just
       - run: just lint-device --release
       - run: just build-device ${{ matrix.board }}
-      - name: Upload firmware artifact
-        uses: actions/upload-artifact@v4
+      - uses: actions/upload-artifact@v4
         with:
-          name: frostsnap-v2-${{ github.run_id }}.elf
+          name: frostsnap-${{ github.sha }}.elf
           path: target/riscv32imc-unknown-none-elf/release/v2
           if-no-files-found: error
 
+  generate-bridge-files:
+    name: Generate Flutter Rust Bridge FFI
+    needs: build-rust-deps
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: subosito/flutter-action@v2
+        with:
+          channel: stable
+          cache: true
+
+      # Restore pub cache
+      - uses: actions/cache@v4
+        with:
+          path: ~/.pub-cache/
+          key: ${{ runner.os }}-pub-cache-${{ hashFiles('frostsnapp/pubspec.lock') }}-v1
+          restore-keys: |
+            ${{ runner.os }}-pub-cache-
+
+      # Restore cargo cache
+      - uses: actions/cache@v4
+        with:
+          path: ~/.cargo/
+          key: ${{ runner.os }}-cargo-registry-${{ hashFiles('./Cargo.lock') }}-v1
+
+      - uses: dtolnay/rust-toolchain@stable
+
+      - name: Setup | Install libudev
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y --no-install-recommends libudev-dev
+
+      - uses: taiki-e/install-action@just
+      - run: just frostsnapp/install-rust-deps
+      - run: just frostsnapp/install-flutter-deps
+      - run: just gen
+
+      - name: Upload Bridge Generated Files
+        uses: actions/upload-artifact@v4
+        with:
+          name: bridge-generated-files
+          path: ./frostsnapp
```

Comment:
What's the thinking behind making the bridge files an artifact -- also why is the whole frostsnapp directory here rather than just the `bridge_` files. I think you can just use the cache again and use this glob:

```
frostsnapp/lib/bridge_*.dart
frostsnapp/native/src/bridge_*.rs
```

and cache it on the contents of `frostsnap/native/src/*`